runtime/engine/src/main/java/io/aklivity/zilla/runtime/engine/internal/registry/WatcherTask.java

```diff
@@ -20,22 +29,65 @@
 public abstract class WatcherTask implements Callable<Void>, Closeable
 {
     private final MessageDigest md5;
-    protected final ScheduledExecutorService executor;
+    private final URL rootConfigURL;
 
+    protected final ScheduledExecutorService executor;
     protected final BiFunction<URL, String, NamespaceConfig> changeListener;
 
     protected WatcherTask(
+        URL rootConfigURL,
         BiFunction<URL, String, NamespaceConfig> changeListener)
     {
+        this.rootConfigURL = rootConfigURL;
         this.changeListener = changeListener;
         this.md5 = initMessageDigest("MD5");
         this.executor  = Executors.newScheduledThreadPool(2);
     }
 
     public abstract Future<Void> submit();
 
-    public abstract String readURL(
-        URL configURL);
+    public String readURL(
+        String location)
+    {
+        String output = null;
+        try
+        {
+            final URL fileURL = new URL(rootConfigURL, location);
+            if ("http".equals(fileURL.getProtocol()) || "https".equals(fileURL.getProtocol()))
+            {
+                HttpClient client = HttpClient.newBuilder()
+                    .version(HTTP_2)
+                    .followRedirects(NORMAL)
+                    .build();
+
+                HttpRequest request = HttpRequest.newBuilder()
+                    .GET()
+                    .uri(fileURL.toURI())
+                    .build();
+
+                HttpResponse<String> response = client.send(
+                    request,
+                    HttpResponse.BodyHandlers.ofString());
+
+                output = response.body();
+            }
+            else
+            {
+
+                URLConnection connection = fileURL.openConnection();
+                try (InputStream input = connection.getInputStream())
+                {
+                    output = new String(input.readAllBytes(), UTF_8);
+                }
+            }
+        }
+        catch (IOException | URISyntaxException | InterruptedException ex)
+        {
+            output = "";
+        }
+        return output;
+    }
+
```

Comment:
Let's consider this in terms of multiple namespaces, each with relocatable relative resources, leading to potentially multiple concurrent "root config urls".